### PR TITLE
ESPIF: Improvements

### PR DESCRIPTION
**Pull request by vladamatena**
- Hide password
- Detect ESPIF failure reason
- Throw away packet data on out of RAM
- Random intron and static MAC size
- Implement protocol version

This is supposed to work with ESP FW:
https://github.com/vladamatena/uart_nic/tree/protocol_version
Binaries are part of the release:
https://github.com/vladamatena/uart_nic/releases/tag/v0

### CODE REVIEW

**Review by vorner — COMMENTED** [review]
I wonder about one thing. What side should be generating the random intron, for two reasons:

* Can someone trick us by a packet data to _overwrite_ our negotiated intron?
* What happens if the ESP resets?

So I wonder if it should be somehow the other way around. Let the ESP send the hello message, then we send a intron to use and that one starts using it.

File: lib/WUI/espif.cpp [hunk 1]
```diff
@@ -79,13 +80,14 @@ enum MessageType {
...
-static const size_t MAC_DATA_MAX_LEN = 16;
+static const uint32_t SUPPORTED_FW_VERSION = 0;
 
-// Thread stuffs
+// Thread stuff
 static void uart_reader(void const *arg);
 osThreadDef(UartBufferThread, uart_reader, osPriorityBelowNormal, 0, 512);
 
 // NIC state
+static std::atomic<uint32_t> fw_version;
```
[the remark this answers is not in the document]
Comment (vladamatena) [reply lib/WUI/espif.cpp:93]: Yes, there is a way to distinguish. The information is printed into logs. We ditinguish:
- Complete silence - no ESP present (or pretending it is not there)
- Some activity on UART (including read errors) - ESP is present, but talking different baudrate/language
- Device info message read from UART, but the protocol version does not match - ESP is present, running our SW just different version

Regarding the version data length, I was also thinking of a shorter integer. On one hand, one should save everywhere, this is embedded. On the other hand, it is send just once. Possible future split into major. minor, (hw revision) version convinced me to use 4 bytes.

File: lib/WUI/espif.cpp [hunk 2]
```diff
@@ -237,23 +237,28 @@ static void process_link_change(bool link_up, struct netif *netif) {
...
     log_debug(ESPIF, "Received ESP data len: %d", size);
+    esp_detected = true;
 
     static enum ProtocolState {
         Intron,
         MessageType,
         Packet,
         PacketLen,
         PacketData,
+        PacketDataThrowaway,
```
[the remark this answers is not in the document]
Comment (vladamatena) [reply lib/WUI/espif.cpp:251]: `PacketData` walks through a list of `pufs`. If allocation fails, there is no list. Possibly, we still can use some if..else to safe the throwaway state. Still, after thinking about it once more I prefer separate state to keep `PacketData` logic simple.

**Comment by vladamatena** [comment]
New revision changes intron configuration logic. Now intron is configured via a separate message send from STM. The intron is now generated by the STM. Thanks for the idea. I like it as the master can skip intron configuration to trade simplicity for security.

File: lib/WUI/espif.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (vladamatena) [reply lib/WUI/espif.cpp]: Latest revision shrinks FW version to a 16bit integer. Regarding the initial fw version 0, I planned to use it for development and switch to 1 just before public release. (I tested the system with a few versions different from 0 just to be sure it works)

**Review by vorner — APPROVED** [review]

**Pull request merged**